# IN this mini project we are going to use the IMDB dataset and learn our RNN (Recurrent Neural Network) Model on this dataset

In [ ]:
from keras.datasets import imdb

((XT,YT),(Xt,Yt))=imdb.load_data(num_words=10000)

In [2]:
print(len(XT),len(YT))

25000 25000


In [0]:
from keras.preprocessing import sequence
X_train=sequence.pad_sequences(XT,maxlen=500)
X_test=sequence.pad_sequences(Xt,maxlen=500)

In [12]:
print(X_train.shape,Y_train.shape)

(25000, 500) (25000, 500)


# Importing the required libraries and compiling our model

In [14]:
from keras.layers import *
from keras.models import Sequential

model=Sequential()
model.add(Embedding(10000,64))
model.add(SimpleRNN(32))
model.add(Dense(1,activation='sigmoid'))


model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_2 (Embedding)      (None, None, 64)          640000    
_________________________________________________________________
simple_rnn_1 (SimpleRNN)     (None, 32)                3104      
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 33        
Total params: 643,137
Trainable params: 643,137
Non-trainable params: 0
_________________________________________________________________


In [15]:
model.compile(optimizer='adam',loss='binary_crossentropy',metrics=['acc'])

W0621 09:47:26.672921 140633269282688 deprecation_wrapper.py:119] From /usr/local/lib/python3.6/dist-packages/keras/optimizers.py:790: The name tf.train.Optimizer is deprecated. Please use tf.compat.v1.train.Optimizer instead.

W0621 09:47:26.710795 140633269282688 deprecation_wrapper.py:119] From /usr/local/lib/python3.6/dist-packages/keras/backend/tensorflow_backend.py:3376: The name tf.log is deprecated. Please use tf.math.log instead.

W0621 09:47:26.721261 140633269282688 deprecation.py:323] From /usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/nn_impl.py:180: add_dispatch_support.<locals>.wrapper (from tensorflow.python.ops.array_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where


# Training our model for 3 epochs

In [17]:
hist=model.fit(X_train,YT,epochs=3,validation_split=0.2,batch_size=128)

Train on 20000 samples, validate on 5000 samples
Epoch 1/3
20000/20000 [==============================] - 34s 2ms/step - loss: 0.5091 - acc: 0.7466 - val_loss: 0.4037 - val_acc: 0.8250
Epoch 2/3
20000/20000 [==============================] - 35s 2ms/step - loss: 0.2859 - acc: 0.8860 - val_loss: 0.3581 - val_acc: 0.8492
Epoch 3/3
20000/20000 [==============================] - 35s 2ms/step - loss: 0.1792 - acc: 0.9356 - val_loss: 0.3622 - val_acc: 0.8632


# Evaluating our model on the Test Dataset

In [19]:
model.evaluate(X_test,Yt)

25000/25000 [==============================] - 27s 1ms/step


[0.38173155264377595, 0.8496]

# We got an accuracy of around 85% on this naive RNN model. This shows how much powerful RNN's can be. Note I picked the top 10,000 words in the dataset.